features/theme.feature

```diff
@@ -41,6 +41,19 @@ Feature: Writing themes
     And I should see "I'm in the project." in "_site/index.html"
     And I should see "<span class=\"sample\">include.html from test-theme</span>" in "_site/index.html"
 
+  Scenario: A theme with data
+    Given I have a configuration file with "theme" set to "test-theme"
+    And I have an _data directory
+    And I have an "_data/greetings.yml" file with content:
+      """
+      foo: "Hello! I’m foo. And who are you?"
+      """
+    And I have an "index.html" page that contains "{{ site.data.greetings.foo }}"
+    When I run jekyll build
+    Then I should get a zero exit status
+    And the _site directory should exist
+    And I should see "Hello! I’m foo. And who are you?" in "_site/index.html"
```

Comment:
This test is unsatisfactory.
You have the same data file with same content at source and in the theme.
How are you asserting the case that the result you've got is from the file in theme and not from the file at source.

Additionally, you should prove that theme data has been overridden by adding a negative assertion:
```gherkin
And I should not see ...
```

Reply:
Where are you seeing he has the same file at source? The point is that he's testing whether data from the theme is picked up.￼

Perhaps you could be a little more clear.￼￼

Reply:
Sorry folks, I’m currently working on provding some tests. Not quite there.

And to get an idea on how this works I just started with the scenario as suggested by @ashmaroli. So, yes, the current addition only tests if the data from the theme was read. But as I couldn’t find how to provide those scenarios that @parkr suggested within the `theme.feature` I‘m actually reading through the `TestThemeAssetsReader` to understand how I can create those tests...

Reply:
@mgerzabek I'm getting mixed signals here.
You're referring to *scenarios*, *features* and `TestThemeAssetsReader` in the same breath. The first two, and the latter belong to two different types of testing interfaces.
*Scenarios*, *features*, etc are terminologies related to Cucumber tests used for *behavior-driven-development*.
`TestThemeAssetsReader` is part of unit-test suite used for *test-driven-development*.

Which aspect are you planning on tackling first..?

Reply:
Thank you @ashmaroli. As I said, I’m completely new to testing with Ruby so don’t know anything. Hence my pledge for some links on the how to and my relatively confused first steps…

As you’ve seen I was able to add a _scenario_ to `theme.feature`. I ran the tests on command line with `./script/test` and got a result that looked okay to me… nothing broken, right? But I’m completely overwhelmed on getting data into these kind of tests other than stating them in sentences… so any hints warmly welcomed!

`TestThemeAssetsReader` is a unit test, a concept that I’m rather familiar with from Java and PHP. I have no clue about setting up new test so I’m just reading the other tests, which is helpful. ATM I have a sceleton which serves as kind of a working plan for the missing pieces.

```ruby
# frozen_string_literal: true

require "helper"

class TestThemeDataReader < JekyllUnitTest

  context "site without a theme" do
    setup do
      @site = fixture_site("theme" => nil)
      @reader = Reader.new(@site)
      @reader.read_data
      assert @site.data.greetings
    end

    should "should read data from source" do

    end
  end

  context "site with a theme without _data" do
    should "should read data from source" do
      site = fixture_site("theme" => "test-theme-skinny")
    end
  end

  context "site with a theme with empty _data directory" do
    should "should read data from source" do
      site = fixture_site("theme" => "test-theme")
    end
  end

  context "site with a theme with data at root of _data" do
    should "should sophistical merge keys" do
      site = fixture_site("theme" => "test-theme")
    end
  end

  context "site with a theme with data at root of _data and in a subdirectory" do
    should "should sophistical merge keys" do
      site = fixture_site("theme" => "test-theme")
    end
  end
end

```

But I wonder how I would setup a **site** where I could overwrite the data from the **theme**. I found a `test/source` directory at the same level as `test/fixture`. This folder contains a `_data` directory. Can I put data files for unit tests there?

Again, any hints warmly welcomed…